src/init.cpp

```diff
@@ -1745,6 +1747,12 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         StartTorControl(onion_service_target);
     }
 
+    if (connOptions.bind_on_any) {
+        // Only add all IP addresses of the machine if we would be listening on
+        // any address - 0.0.0.0 (IPv4) and :: (IPv6).
+        Discover();
```

Comment:
2aee1b92 verified that without the change in this commit, the accompanying functional test fails at the expected assertion
```
File "~/bitcoin/test/functional/feature_bind_port_discover.py", line 74, in run_test
    assert local['address'] != ADDR2
AssertionError
```